src/transformers/models/xcodec2/modular_xcodec2.py

```diff
@@ -0,0 +1,1653 @@
+
+        Args:
+            audio (`torch.Tensor` of shape `(batch_size, 1, sequence_length)`):
+                Input audio waveform.
+            audio_spectrogram (`torch.Tensor` of shape `(batch_size, mel_bins, time_steps)`):
+                Input audio mel spectrogram for semantic encoding.
+            padding_mask (`torch.Tensor` of shape `(batch_size, 1, sequence_length)`):
+                Padding mask used to pad `audio`.
+            return_dict (`bool`, *optional*):
+                Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+
+        Returns:
+            `audio_codes` of shape `[batch_size, 1, frames]`, the discrete encoded codes for the input audio waveform.
+            `quantized_representation` of shape `[batch_size, hidden_size, frames]`, the continuous quantized
+                representation after quantization.
+            `codes_padding_mask` of shape `[batch_size, 1, frames]`, downsampled `padding_mask` for indicating valid
+                audio codes in `audio_codes`.
+        """
+        return_dict = return_dict if return_dict is not None else self.config.return_dict
+
+        # 1) Semantic embedding: 16th layer of pretrained model: https://huggingface.co/HKUSTAudio/xcodec2/blob/main/modeling_xcodec2.py#L64
+        semantic_output = self.semantic_model(audio_spectrogram, output_hidden_states=True)
+        semantic_hidden_16 = semantic_output.hidden_states[16]
+        semantic_hidden_16 = semantic_hidden_16.transpose(1, 2)
+        semantic_encoded = self.semantic_encoder(semantic_hidden_16)
+
+        # 2) Get acoustic embedding
+        vq_emb = self.acoustic_encoder(audio)
+        vq_emb = vq_emb.transpose(1, 2)
+
+        # 3) Concat embeddings and apply final layers
+        if vq_emb.shape[-1] != semantic_encoded.shape[-1]:
+            min_len = min(vq_emb.shape[-1], semantic_encoded.shape[-1])
+            vq_emb = vq_emb[:, :, :min_len]
+            semantic_encoded = semantic_encoded[:, :, :min_len]
+        concat_emb = torch.cat([semantic_encoded, vq_emb], dim=1)
+        concat_emb = self.fc_prior(concat_emb.transpose(1, 2)).transpose(1, 2)
+
+        # 4) Get codes for decoder
+        quantized_representation, audio_codes = self.decoder.quantize(concat_emb)
+
+        # If provided, compute corresponding padding mask for audio codes
+        codes_padding_mask = None
+        if padding_mask is not None:
+            # Expected token length, as in: https://github.com/zhenye234/X-Codec-2.0/blob/ccbbf340ff143dfa6a0ea7cd61ec34a8ba2f1c3d/inference_save_code.py#L89
+            audio_length = padding_mask.sum(dim=-1, keepdim=True).cpu()
+            token_length = audio_length // self.hop_length
+            codes_padding_mask = torch.zeros(audio_codes.shape, dtype=padding_mask.dtype)
+            idx = torch.arange(audio_codes.shape[-1]).view(1, -1)
+            codes_padding_mask = (idx < token_length).to(padding_mask.dtype).to(padding_mask.device)
```

Comment:
Using provided padding mask to compute corresponding padding mask in tokens "space"